GPU/Common/FramebufferManagerCommon.cpp

```diff
@@ -2395,33 +2396,31 @@ bool FramebufferManagerCommon::GetOutputFramebuffer(GPUDebugBuffer &buffer) {
 	return retval;
 }
 
-// This function takes an already correctly-sized framebuffer and packs it into RAM.
+// This function takes an already correctly-sized framebuffer and reads it into emulated PSP VRAM.
 // Does not need to account for scaling.
+//
 // Color conversion is currently done on CPU but should theoretically be done on GPU.
 // (Except using the GPU might cause problems because of various implementations'
 // dithering behavior and games that expect exact colors like Danganronpa, so we
 // can't entirely be rid of the CPU path.) -- unknown
-void FramebufferManagerCommon::PackFramebufferSync_(VirtualFramebuffer *vfb, int x, int y, int w, int h) {
-	if (!vfb->fbo) {
-		ERROR_LOG_REPORT_ONCE(vfbfbozero, SCEGE, "PackFramebufferSync_: vfb->fbo == 0");
-		return;
-	}
-
+void FramebufferManagerCommon::PackFramebufferSync(VirtualFramebuffer *vfb, int x, int y, int w, int h, RasterChannel channel) {
 	if (w <= 0 || h <= 0) {
-		ERROR_LOG(G3D, "Bad inputs to PackFramebufferSync_: %d %d %d %d", x, y, w, h);
+		ERROR_LOG(G3D, "Bad inputs to PackFramebufferSync: %d %d %d %d", x, y, w, h);
 		return;
 	}
 
 	const u32 fb_address = vfb->fb_address & 0x3FFFFFFF;
 
-	Draw::DataFormat destFormat = GEFormatToThin3D(vfb->fb_format);
+	Draw::DataFormat destFormat = channel == RASTER_COLOR ? GEFormatToThin3D(vfb->fb_format) : GEFormatToThin3D(GE_FORMAT_DEPTH16);
 	const int dstBpp = (int)DataFormatSizeInBytes(destFormat);
 
-	const int dstByteOffset = (y * vfb->fb_stride + x) * dstBpp;
-	const int dstSize = (h * vfb->fb_stride + w - 1) * dstBpp;
+	int stride = channel == RASTER_COLOR ? vfb->fb_stride : vfb->z_stride;
+
+	const int dstByteOffset = (y * stride + x) * dstBpp;
+	const int dstSize = (h * stride + w - 1) * dstBpp;  // TODO: What's that "w - 1" doing here?
```

Comment:
Because the last line isn't as long as the others, when w < stride which is typical.

For example, consider a block transfer from the stride gap.  Stride is 512 (because the larger framebuffer is 512), but width is 32 and height is 272.  If we calculate the total bytes based on 512, we will go over the end of the region by 480, and this might even be outside valid VRAM.

At one point I thought I made it handle the stride gap (if necessary when we didn't estimate for its width so clearly didn't have those pixels) using a temp buffer, but maybe that's for something else or has been refactored away.

-[Unknown]

Reply:
Ah, of course. Will update my comment.